### DISCUSSION TITLE
@rematch/immer - why the different pattern?

### DISCUSSION THREAD

**Discussion by s-h-a-d-o-w** [rematch/rematch#932] [Q&A] [answered]
It seems to me that if the plugin worked the same way that immer and redux toolkit do it (no return required), it would have both reduced boiler plate as well as the behavior that people would expect without reading the docs.

Which is why I'm curious why you decided to go a different route. Maybe there's some upside that I'm missing. :)

**Comment by semoal** [comment, marked as answer]
Hello Andreas, Rematch has been published for the first time 4 years ago... 4 years ago introducing Immer by default to our users wasn't a good idea, Immer wasn't enough stable yet and also we weren't so much experienced with Redux.
Adding @rematch/immer to a Rematch project it's as easy as adding it to the `plugins` array so I think trying to ship as little as code as possible to @rematch/core was a good idea (that I still think it's a good idea)

Redux Toolkit was created 2 years ago, really influenced by solutions like Rematch and brings a new and cooler Redux, which I highly recommend if it fits your use case, if you like Rematch, just use Rematch (I'll keep maintaining Rematch as long as I can)

Hope this answers your question :) Feel free to ask anything else!

**Reply by s-h-a-d-o-w** [reply]
Thanks for the quick and elaborate response but I'm afraid you misunderstood my question. Because it wasn't about why you don't include immer by default (it seems like that's what your answer is about) but purely about the state manipulation pattern in the immer plugin itself.

To emphasize the key part of my question: Immer/RTK don't require returning the state object but your plugin does. Why?
(Also - I'm sorry if that's a decision made like 2 years ago and nobody can remember. :D But if that's the case, maybe it's worthwhile revisiting whether it should be adjusted to be consistent with these other libraries?)

**Reply by semoal** [reply]
Ahp sorry! hehe. We support not returning the state object when using Immer, but to be honest, never thought about that until you mentioned it.

To make sure I committed a change on our Immer tests, I removed the return and the suite passed correctly :)

https://github.com/rematch/rematch/commit/9803b507e08c0668d0c0b5d53c991b1dc0552cd3 

I'll update all the documentation sections where we return the state when using Immer right now

**Reply by s-h-a-d-o-w** [reply]
That's so awesome, now I'll definitely give rematch a shot at the next opportunity. Thanks! :)

**Comment by s-h-a-d-o-w** [comment]
I just checked your updates regarding this and saw that this text in the docs still needs to be removed: 

> Immer doesn't require that you return the next state from a reducer, but @rematch/immer plugin expects you to do it! Your reducers must always return the next state. Otherwise, you will reset your model's state.

**Reply by semoal** [reply]
You're right mate, thx for pointing it out. Totally missed that section also the 2nd line is not true anymore since Immer works out of the box with primitive values.
ps: fixed =)
